pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1322 @@
+			if state.is_active() {
+				Self::update_active(collator.clone(), state.total);
+			}
+			<CollatorState<T>>::insert(&collator, state);
+			Self::deposit_event(Event::CollatorBondedMore(collator, before, after));
+			Ok(().into())
+		}
+		/// Bond less for collator candidates
+		#[pallet::weight(0)]
+		pub fn candidate_bond_less(origin: OriginFor<T>, less: BalanceOf<T>) -> DispatchResultWithPostInfo {
+			let collator = ensure_signed(origin)?;
+			let mut state = <CollatorState<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
+			ensure!(!state.is_leaving(), Error::<T>::CannotActivateIfLeaving);
+			let before = state.bond;
+			let after = state.bond_less(less).ok_or(Error::<T>::Underflow)?;
+			ensure!(after >= T::MinCollatorCandidateStk::get(), Error::<T>::ValBondBelowMin);
+			T::Currency::unreserve(&collator, less);
+			if state.is_active() {
+				Self::update_active(collator.clone(), state.total);
+			}
+			<CollatorState<T>>::insert(&collator, state);
+			Self::deposit_event(Event::CollatorBondedLess(collator, before, after));
+			Ok(().into())
+		}
+		/// If caller is not a delegator, then join the set of delegators
+		/// If caller is a delegator, then makes delegation to change their
+		/// delegation state
+		#[pallet::weight(0)]
+		pub fn delegate(
+			origin: OriginFor<T>,
+			collator: T::AccountId,
+			amount: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let acc = ensure_signed(origin)?;
+			if let Some(mut delegator) = <DelegatorState<T>>::get(&acc) {
+				// delegation after first
+				ensure!(amount >= T::MinDelegation::get(), Error::<T>::DelegationBelowMin);
+				ensure!(
+					(delegator.delegations.0.len() as u32) < T::MaxCollatorsPerDelegator::get(),
+					Error::<T>::ExceedMaxCollatorsPerNom
+				);
+				let mut state = <CollatorState<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
+				ensure!(
+					delegator.add_delegation(Bond {
+						owner: collator.clone(),
+						amount
+					}),
+					Error::<T>::AlreadyDelegatedCollator
+				);
+				let delegation = Bond {
+					owner: acc.clone(),
+					amount,
+				};
+				// TODO: How to handle: Kick lowest bonded delegators or keep as is?
+				ensure!(
+					(state.delegators.0.len() as u32) < T::MaxDelegatorsPerCollator::get(),
+					Error::<T>::TooManyDelegators
+				);
+				ensure!(state.delegators.insert(delegation), Error::<T>::DelegatorExists);
+				T::Currency::reserve(&acc, amount)?;
+				let new_total = state.total + amount;
+				if state.is_active() {
+					Self::update_active(collator.clone(), new_total);
+				}
+
+				// Update states
+				let (total_collators, total_delegators) = <Total<T>>::get();
+				<Total<T>>::put((total_collators, total_delegators.saturating_add(amount)));
+				state.total = new_total;
+				<CollatorState<T>>::insert(&collator, state);
+				<DelegatorState<T>>::insert(&acc, delegator);
+				Self::deposit_event(Event::Delegation(acc, amount, collator, new_total));
+			} else {
+				// first delegation
```

Comment:
such a big `if`/`else` statement will be hard to weigh. Might be worth just making a separate `first_delegate` extrinsic or similar.